### PR TITLE
fix(core): don’t recreate `TemplateRef` when used as a reference.

### LINKED ISSUE #14873: ngIf else statement always re-assign template reference variable when component's change detection occur
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

In my angular 4 hybrid web app, i tried to upgrade from 4.0.0-beta.8 to 4.0.0-rc.1 (also i tried 4.0.0-rc2). When upgraded, *ngIf else statement make component which is inside ng-template reload. 

**Expected behavior**

template reference variable should be re-assign when *ngIf condition's variable changed.

**Minimal reproduction of the problem with instructions**

http://plnkr.co/edit/dXcEXDcYKDnyvNs0DwBZ

```
@Component({
  selector: 'angular-test',
  template: `
    <div *ngIf="test; else postBlock;">
      test
    </div>
    
    <ng-template #postBlock>
      <angular-test2></angular-test2>
</ng-template>
    `
})
export class TestComponent implements OnInit {
  private test:boolean;

  ngOnInit(): void {

  }
}


@Component({
  selector: 'angular-test2',
  template: `
tsetse
    `
})
export class Test2Component implements OnInit {
  private test:boolean;

  ngOnInit(): void {
    console.log('test2');
  }
}

```

```
<angular-test></angular-test> 
```

angular-test component's ng-template #postBlock makes angular-test2 component reload (console.log always occur when screen touch or timer or some component has changed) 

**What is the motivation / use case for changing the behavior?**

In my app, post-detail component load post object from http request and then, 
`<div *ngIf="!post; else postBlock"></div> <ng-template #postBlock> ~~ <post-comment></post-comment> </ng-template> `
code render #postBlock's inner html. 
In post-comment component, it request http to getting comments in ngOnInit(). 
When i upgraded from beta.8 to rc.1/rc.2, post-comment's ngOnInit function call is infinitely occured. 

**Please tell us about your environment:**

@angular/cli: 1.0.0-rc.0
node: 7.5.0
os: darwin x64
@angular/animations: 4.0.0-rc.2
@angular/cli: 1.0.0-rc.0
@angular/common: 4.0.0-rc.2
@angular/compiler: 4.0.0-rc.2
@angular/compiler-cli: 4.0.0-rc.2
@angular/core: 4.0.0-rc.2
@angular/forms: 4.0.0-rc.2
@angular/http: 4.0.0-rc.2
@angular/material: 2.0.0-beta.2
@angular/platform-browser: 4.0.0-rc.2
@angular/platform-browser-dynamic: 4.0.0-rc.2
@angular/platform-server: 4.0.0-rc.2
@angular/router: 4.0.0-rc.2
@ngtools/webpack: 1.2.11

* **Angular version:** 4.0.0-rc.1 or 4.0.0-rc.2

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript 2.2.1

* **Node (for AoT issues):** `node --version` =   7.5.0

### PR DESCRIPTION
**Pull request by tbosch**
This was a regression introduced in v4 rc.0.

Fixes #14873

### CODE REVIEW

File: modules/@angular/core/test/linker/regression_integration_spec.ts [hunk 1]
```diff
@@ -222,6 +222,29 @@ function declareTests({useJit}: {useJit: boolean}) {
       const txtNode = ctx.componentInstance.renderer.createText('test');
       expect(txtNode).toHaveText('test');
     });
+
+    it('should not recreate TemplateRef references during dirty checking', () => {
+      @Component({template: '<div [someDir]="someRef"></div><template #someRef></template>'})
```
Comment (leo6104) [inline modules/@angular/core/test/linker/regression_integration_spec.ts]: @tbosch Thanks to pull request for #14873  issues. I think ng-template tag is better than template tag for angular 4. Is that intentional for angular 2 testing compatibility?

Comment (vicb) [inline modules/@angular/core/test/linker/regression_integration_spec.ts]: Yep, should be `ng-template`

File: modules/@angular/core/test/linker/regression_integration_spec.ts [hunk 2]
```diff
@@ -222,6 +222,29 @@ function declareTests({useJit}: {useJit: boolean}) {
...
+
+      @Directive({selector: '[someDir]'})
+      class MyDir {
+        @Input('someDir') template: TemplateRef<any>;
+      }
+
+      TestBed.configureTestingModule({declarations: [MyComp, MyDir]});
+      const ctx = TestBed.createComponent(MyComp);
+      const dir = <MyDir>ctx.debugElement.children[0].injector.get(MyDir);
+      expect(dir.template).toBeUndefined();
```
Comment (vicb) [inline modules/@angular/core/test/linker/regression_integration_spec.ts]: nits:
- chain calls for TestBed,
- I think we should be use `.query(By.directive())` when possible to make tests more explicit.

File: modules/@angular/core/src/view/query.ts [hunk 3]
```diff
@@ -180,10 +179,10 @@ export function getQueryValue(
         value = new ElementRef(asElementData(view, nodeDef.index).renderElement);
         break;
       case QueryValueType.TemplateRef:
-        value = createTemplateRef(view, nodeDef);
+        value = asElementData(view, nodeDef.index).template;
         break;
       case QueryValueType.ViewContainerRef:
-        value = createViewContainerRef(view, nodeDef);
+        value = asElementData(view, nodeDef.index).viewContainer;
```
Comment (vicb) [inline modules/@angular/core/src/view/query.ts]: test ?

File: modules/@angular/core/src/view/refs.ts [hunk 4]
```diff
@@ -84,15 +84,17 @@ class ComponentRef_ extends ComponentRef<any> {
...
 }
 
-class ViewContainerRef_ implements ViewContainerRef {
-  private _data: ElementData;
-  constructor(private _view: ViewData, private _elDef: NodeDef) {
-    this._data = asElementData(_view, _elDef.index);
-  }
+class ViewContainerRef_ implements ViewContainerData {
+  /**
+   * @internal
```
Comment (vicb) [inline modules/@angular/core/src/view/refs.ts]: needed ?

File: modules/@angular/core/src/view/refs.ts [hunk 5]
```diff
@@ -84,15 +84,17 @@ class ComponentRef_ extends ComponentRef<any> {
...
+   */
+  _embeddedViews: ViewData[] = [];
+  constructor(private _view: ViewData, private _elDef: NodeDef, private _data: ElementData) {}
```
Comment (vicb) [inline modules/@angular/core/src/view/refs.ts]: seems like we only need to keep a ref to the render el

File: modules/@angular/core/test/linker/regression_integration_spec.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/core/test/linker/regression_integration_spec.ts]: Done

File: modules/@angular/core/test/linker/regression_integration_spec.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/core/test/linker/regression_integration_spec.ts]: Done

File: modules/@angular/core/src/view/refs.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/core/src/view/refs.ts]: Decided to keep it.

File: modules/@angular/core/src/view/query.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/core/src/view/query.ts]: Done

**Review by vicb — APPROVED** [review]

**Review by mhevery — APPROVED** [review]

**Pull request merged**